redhat/rdma-core.spec

```diff
@@ -108,6 +108,7 @@ RDMA core development libraries and headers.
 
 %package -n infiniband-diags
 Summary: InfiniBand Diagnostic Tools
+Requires: libibumad%{?_isa} = %{version}-%{release}
```

Comment:
I don't understand. Why doesn't rpm build pickup the right shlib dependenciess automtically? 

We certainly don't have a = dependency like this.

Why would we ever need a specific package version requirement when we use symbol versions properly?

At worst should should just be >= some version at the start of the rdma-core era, but I think you'd need to justfiy that with some old libibumad that forces it.

Reply:
> I don't understand. Why doesn't rpm build pickup the right shlib dependenciess automtically?

rpm build do pickup the right shlib dependency. Because multiple versions libibumad provide identical list capabilities of provides, mixed version packages may be installed when multiple versions of packages are available in yum/dnf repositories. It is terrible for package building and customer support, when mixed version packages installed.

$ rpm -qpR infiniband-diags-28.0-1.fc32.x86_64.rpm | grep umad
libibumad.so.3()(64bit)
libibumad.so.3(IBUMAD_1.0)(64bit)
libibumad.so.3(IBUMAD_1.1)(64bit)

> Why would we ever need a specific package version requirement when we use symbol versions properly?
 
Please see my inline comment.

> At worst should should just be >= some version at the start of the rdma-core era, but I think you'd need to justfiy that with some old libibumad that forces it.

No, ">=' is helpless when multiple versions libibumad available in yum/dnf repository.

Reply:
The shlib dependencies are supposed to be sufficient. If there is a historical problem with them then you need to identify exactly what it is and we can move forward properly. This = dependency is not correct.

Reply:
It is standard convention in rpm packaging that there be a hard = requirement between subpackages of a single source package.  Usually this is done in each subpackage by making it require the parent package be the same version.  But when you have a cross dependency like this, it makes sense to do it there as well.  That way if you do dnf upgrade rdma-core on the command line, the entire package set gets updated.  We rely on automatic library dependencies for external libraries, but not for same source package libraries.  There are reasons for that (although they are old and well established reasons that I have mostly long since forgotten).  As an example, a new build of the package may have changed the compile parameters in a way that materially effects the library without changing the symbol versions.

In any case, the typical situation is that *all* of the subpackages would have:

Requires: rdma-core${isa} = %{version}-%{release}

Then if a user tries to upgrade a single package like libibumad, it will trigger pulling in the updated rdma-core, which will then trigger all of the other subpackages to be upgraded too.  If we had set that up properly in the first place, this libibumad cross dependency would be redundant.  So it might be worth just fixing it the standard way instead of this way.

Reply:
It certainly makes more sense to do this consistently everywhere if it is RPM policy. Otherwise the shlib packages are supposed to be versioned soley by symbol versions and no RPM version restriction should ever be needed. If this is not true it needs to be explained and delt with.

Reply:
https://docs.fedoraproject.org/en-US/packaging-guidelines/#_requiring_base_package

The official policy is that if there are dependencies between base and subpackages, then yes, an explicit requires of the base package in the subpackage on the entire release and version is expected.

Reply:
I don't think that language applies here. The proposed depends is not a depends on the base but a depends on a different sub package (ie a library) which must be fully and correctly versioned using only the symver mechanism or things are fundamentally broken.

Reply:
https://docs.fedoraproject.org/en-US/packaging-guidelines/#_explicit_requires

I believe we fall into this category in this instance, in which case we should have the core package requires.  The soname dependency fails in the specific instance that:

1) library is extended in backword compatible manner
2) application builds against later library and uses newer extensions
3) application is then installed on system with downrev library missing extensions

I saw the original Red Hat bug report that made Honggang write this patch, and I think it fell into that category, but I'm going from memory so I could be wrong.  In any case, we often extend our libraries without bumping the major sonumber, so it seems entirely appropriate to do the specific requires on the core package in all subpackages to make sure that our applications and our libraries stay in sync.

Reply:
Like I keep saying, if symbol versions have been used wrongly in the past, then we need to actually know the actual details of that and can form a repair strategy.

Broken symver could effect any package, using libibmad, while adding an explicit requires here will only fix the one case of the package inside rdma-core. We need a proper fix, if that really is the problem.

Reply:
I will drop this commit.  We will create a fedora/RHEL specific patch for our internal Fedora/RHEL spec file.